File: spark/tools/shell.py
"""Shell command execution tool."""

import subprocess
from typing import Any

from spark.tool import tool


@tool
def run_command(command: str, timeout: int = 120000) -> str:
    """
    Execute a shell command and return its output.

    Args:
        command: The shell command to execute
        timeout: Timeout in milliseconds (default: 120000ms = 2 minutes)

    Returns:
        Command output (stdout and stderr), or error message
    """
    if not command or not command.strip():
        return "Error: Empty command"

    try:
        result = subprocess.run(
            command,
            shell=True,
            capture_output=True,
            text=True,
            timeout=timeout / 1000,  # Convert to seconds
        )

        output_parts = []

        if result.stdout:
            output_parts.append(result.stdout)

        if result.stderr:
            output_parts.append(f"STDERR:\n{result.stderr}")

        if result.returncode != 0:
            output_parts.append(f"Exit code: {result.returncode}")

        output = '\n'.join(output_parts).strip()

        return output if output else f"Command completed with exit code {result.returncode}"

    except subprocess.TimeoutExpired:
        return f"Error: Command timed out after {timeout}ms"
    except Exception as e:
        return f"Error: {str(e)}"

File: src/tools/shell.ts
import { spawn } from "node:child_process";
import type { Tool } from "./index.js";
import { SafetyChecker } from "../safety.js";

let projectDir = process.cwd();
let safetyChecker = new SafetyChecker({ projectRoot: projectDir });

export function setShellProjectDir(dir: string): void {
  projectDir = dir;
  safetyChecker = new SafetyChecker({ projectRoot: dir });
}

const DEFAULT_TIMEOUT = 120_000;

/** Maximum characters collected from stderr before truncating (prevents output explosion). */
const MAX_STDERR_CHARS = 10_000;

const runCommand: Tool = {
  name: "run_command",
  description:
    "Execute a shell command with real-time streaming output. Uses cmd /c on Windows, /bin/bash -c on other platforms.",
  parameters: {
    command: { type: "string", description: "Shell command to execute" },
    timeout: {
      type: "number",
      description: "Timeout in milliseconds (default: 120000)",
    },
  },
  required: ["command"],
  requiresConfirmation: true,
  async execute(args) {
    const command = String(args.command);
    const timeout = args.timeout ? Number(args.timeout) : DEFAULT_TIMEOUT;

    try {
      safetyChecker.checkCommand(command);
    } catch (err) {
      return err instanceof Error ? err.message : String(err);
    }

    return new Promise((resolvePromise) => {
      const isWindows = process.platform === "win32";

      let shell: string;
      let shellArgs: string[];

      if (isWindows) {
        const escaped = command.replace(/%/g, "%%").replace(/"/g, '\\"');
        shell = "cmd";
        shellArgs = ["/d", "/c", escaped];
      } else {
        shell = "/bin/bash";
        shellArgs = ["-c", command];
      }

      const child = spawn(shell, shellArgs, {
        cwd: projectDir,
        timeout,
        windowsHide: true,
      });

      let stdout = "";
      let stderr = "";

      child.stdout?.on("data", (data: Buffer) => {
        stdout += data.toString();
      });

      child.stderr?.on("data", (data: Buffer) => {
        const text = data.toString();
        // Limit stderr collection to prevent output explosion
        if (stderr.length < MAX_STDERR_CHARS) {
          stderr += text;
        }
      });

      child.on("error", (err) => {
        if ((err as NodeJS.ErrnoException).code === "ERR_CHILD_PROCESS_TIMEOUT") {
          resolvePromise(`Command timed out after ${timeout}ms`);
          return;
        }
        resolvePromise(`Error: ${err.message}`);
      });

      child.on("close", (code) => {
        if (code !== null && code !== 0) {
          let output = "";
          if (stderr) {
            output += stderr;
            if (stderr.length >= MAX_STDERR_CHARS) {
              output += "\n... (stderr truncated)";
            }
          }
          if (stdout) {
            output += (output ? "\n" : "") + stdout;
          }
          output += `\nCommand exited with code ${code}`;
          resolvePromise(output);
          return;
        }

        let output = "";
        if (stdout) output += stdout;
        if (stderr) {
          output += (output ? "\n" : "") + stderr;
        }
        resolvePromise(output || "(no output)");
      });
    });
  },
};

export const shellTools: Tool[] = [runCommand];
